File: Tests/ImageCompression/Utils/DCTMultiThreadTest.java
package ImageCompression.Utils;

import ImageCompression.Utils.Functions.DCTMultiThread;
import org.junit.Before;
import org.junit.Test;

import java.util.Random;

import static org.junit.Assert.*;

public class DCTMultiThreadTest {
    short[][] data,res;
    int size= DCTMultiThread.SIZEOFBLOCK;
    @Before
    public void setUp() throws Exception {
        data=new short[size][size];
        for(int i=0;i<size;i++){
            for (int j=0;j<size;j++){
                data[i][j]=(short)(i*j);
            }
        }
    }

    @Test
    public void TestDctEquals() throws Exception {
        res=DCTMultiThread.directDCT(data);
        res=DCTMultiThread.reverseDCT(res);

//        assertArrayEquals(res,data);
        assertArrayInRange(data,res,0);
    }
    @Test
    public void TestDCTinRange1() throws Exception {
        res=DCTMultiThread.directDCT(data);
        res=DCTMultiThread.reverseDCT(res);

//        assertArrayEquals(res,data);
        assertArrayInRange(data,res,1);
    }
    @Test
    public void MultiDCT() throws Exception{
        Random random=new Random();
        for(int i=0;i<10;i++){
            for(int x=0;x<size;x++){
                for(int y=0;y<size;y++){
                    data[x][y]=(short)random.nextInt(255);
                }
            }
            res=DCTMultiThread.directDCT(data);
            res=DCTMultiThread.reverseDCT(res);

            assertArrayInRange(data,res,2);
        }
    }

    private void assertArrayInRange(short[][] a,short[][] b,int delta){
        assertEquals(a.length,b.length);
        assertTrue(a.length>0);
        assertEquals(a[0].length,b[0].length);

        for(int i=0;i<a.length;i++){
            for(int j=0;j<a[0].length;j++){
                boolean isEqual=false;
                for(int d=-delta;d<=delta;d++){
                    if(a[i][j]==(b[i][j]+d))
                        isEqual=true;
                }
                assertTrue("in ["+i+"]["+j+"] value "+a[i][j]+"!="+b[i][j],isEqual);
            }
        }
    }

}

File: Tests_tmp/ImageCompressionLib/Utils/DCTMultiThreadTest.kt
package ImageCompressionLib.Utils

import ImageCompressionLib.Constants.SIZEOFBLOCK
import ImageCompressionLib.Containers.Matrix.ShortMatrix
import ImageCompressionLib.Containers.Type.Size
import ImageCompressionLib.Utils.Functions.DCTMultiThread
import org.junit.Before
import org.junit.Test

import java.util.Random

import org.junit.Assert.*
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import kotlin.test.assertFails

@RunWith(Parameterized::class)
class DCTMultiThreadTest(val size:Size) {
    lateinit var data:ShortMatrix
    @Before
    fun setUp() {
        data = ShortMatrix(size){i, j -> (Math.abs(Random().nextInt())%255).toShort() }
    }

    @Test
    fun TestDctNotEquals() {
        var res = ShortMatrix.valueOf(DCTMultiThread.directDCT(data))
        res = ShortMatrix.valueOf(DCTMultiThread.reverseDCT(data))

        assertFails { data.assertInRange( res, 0)}
    }

    @Test
    fun TestDCTinRange1() {
        val cpy=data.copy()
        val dir = ShortMatrix.valueOf(DCTMultiThread.directDCT(data))
        assertFails { cpy.assertInRange( dir, 0)}
        val rev = ShortMatrix.valueOf(DCTMultiThread.reverseDCT(dir))
        cpy.assertInRange(rev,1)
    }
    @Test
    @Throws(Exception::class)
    fun TestDCTinRange3() {
        val cpy=data.copy()
        val dir = ShortMatrix.valueOf(DCTMultiThread.directDCT(data))
        assertFails { cpy.assertInRange( dir, 0)}
        val rev = ShortMatrix.valueOf(DCTMultiThread.reverseDCT(dir))
        cpy.assertInRange(rev,3)
    }
    companion object{
        @JvmStatic
        @Parameterized.Parameters(name="{0}")
        fun data(): Collection<Array<Size>> {
            return listOf(
                    arrayOf(Size(8,8))
//                    ,arrayOf(Size(7,7))
//                    ,arrayOf(Size(7,9))
            )
        }
    }


}